File: examples/kotlin/examples/ClientSwrNative.kt
package io.ekodb.client.examples

/**
 * ekoDB Kotlin Client - Native SWR Function Example
 *
 * Demonstrates the native SWR (Stale-While-Revalidate) function that
 * encapsulates the entire cache-aside pattern in a single operation.
 */

import io.ekodb.client.EkoDBClient
import io.ekodb.client.functions.FunctionStageConfig
import io.ekodb.client.functions.ParameterDefinition
import io.ekodb.client.functions.UserFunction
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlin.system.measureTimeMillis

private const val BASIC_LABEL = "github_user_native_kt"
private const val AUDIT_LABEL = "product_swr_audit_kt"
private const val PIPELINE_LABEL = "user_enrichment_pipeline_kt"
private const val DYNAMIC_LABEL = "flexible_cache_kt"
private const val AUDIT_COLLECTION = "swr_audit_trail_kt"
private const val ENRICHED_COLLECTION = "enriched_users_kt"

private fun isAlreadyExistsError(e: Exception): Boolean {
    val msg = e.message ?: return false
    return msg.contains("status 409") || msg.contains("already exists")
}

private fun isNotFoundError(e: Exception): Boolean =
    e.message.orEmpty().contains("not found", ignoreCase = true)

private suspend fun saveOrUpdate(client: EkoDBClient, func: UserFunction): String {
    return try {
        client.saveFunction(func)
    } catch (e: Exception) {
        if (isAlreadyExistsError(e)) {
            client.updateFunction(func.label, func)
            println("ℹ️  Function '${func.label}' already existed — updated instead")
            client.getFunction(func.label).id
                ?: throw IllegalStateException("No ID returned for function '${func.label}'")
        } else {
            throw e
        }
    }
}

fun exampleBasicSWR(client: EkoDBClient): String = runBlocking {
    println("\nExample 1: Basic Native SWR")
    println("─".repeat(80))
    println("Single function replaces KvGet → If → HttpRequest → KvSet pipeline")

    // Create function with native SWR function
    val basicSWRScript = UserFunction(
        label = BASIC_LABEL,
        name = "GitHub User Lookup (Native SWR)",
        description = "Fetches GitHub user data with automatic caching using native SWR",
        parameters = mapOf(
            "username" to ParameterDefinition(
                required = true,
                description = "GitHub username to fetch",
                paramType = "String"
            )
        ),
        functions = listOf(
            FunctionStageConfig.SWR(
                cache_key = "github:user:kt:{{username}}",
                ttl = JsonPrimitive("15m"),
                url = "https://api.github.com/users/{{username}}",
                method = "GET",
                headers = mapOf("User-Agent" to "ekoDB-Client"),
                body = null,
                timeout_seconds = null,
                output_field = "user_data",
                collection = null
            )
        ),
        tags = listOf("github", "swr", "native")
    )

    val funcId = saveOrUpdate(client, basicSWRScript)
    println("✓ Created native SWR function: $BASIC_LABEL ($funcId)")

    // First call - cache miss
    println("\nFirst call (cache miss - will fetch from GitHub API):")
    var recordCount1 = 0
    val duration1 = measureTimeMillis {
        val result1 = client.callFunction(BASIC_LABEL, mapOf("username" to JsonPrimitive("torvalds")))
        recordCount1 = result1.records.size
    }
    println("  Response time: ${duration1}ms")
    println("  Records returned: $recordCount1")

    // Second call - cache hit
    println("\nSecond call (cache hit - instant from KV store):")
    val duration2 = measureTimeMillis {
        client.callFunction(BASIC_LABEL, mapOf("username" to JsonPrimitive("torvalds")))
    }
    println("  Response time: ${duration2}ms")
    val speedup = if (duration2 > 0) duration1.toDouble() / duration2.toDouble() else 0.0
    println("  Speedup: ${"%.1f".format(speedup)}x faster 🚀")
    println()

    funcId
}

fun exampleAuditTrail(client: EkoDBClient): String = runBlocking {
    println("\nExample 2: SWR with Built-in Audit Trail")
    println("─".repeat(80))
    println("Optional collection parameter for automatic request logging")

    // Create function with audit trail
    val auditSWRScript = UserFunction(
        label = AUDIT_LABEL,
        name = "Product API with Audit (Native SWR)",
        description = "Caches product data and logs all requests automatically",
        parameters = mapOf(
            "product_id" to ParameterDefinition(
                required = true,
                description = "Product ID to fetch",
                paramType = "String"
            )
        ),
        functions = listOf(
            FunctionStageConfig.SWR(
                cache_key = "product:kt:{{product_id}}",
                ttl = JsonPrimitive("1h"),
                url = "https://fakestoreapi.com/products/{{product_id}}",
                method = "GET",
                headers = null,
                body = null,
                timeout_seconds = null,
                output_field = "product",
                collection = AUDIT_COLLECTION
            )
        ),
        tags = listOf("products", "audit")
    )

    val auditFuncId = saveOrUpdate(client, auditSWRScript)
    println("✓ Created SWR function with audit trail: $AUDIT_LABEL ($auditFuncId)")

    println("\nFetching product (will create audit trail entry):")
    val productResult = client.callFunction(AUDIT_LABEL, mapOf("product_id" to JsonPrimitive("1")))
    println("  ✓ Product fetched and cached")
    println("  ✓ Audit record created in '$AUDIT_COLLECTION' collection")
    println("  Records: ${productResult.records.size}\n")

    auditFuncId
}

fun examplePipelineEnrichment(client: EkoDBClient): String = runBlocking {
    println("\nExample 3: SWR in Multi-Function Pipeline")
    println("─".repeat(80))
    println("Fetch external data → Process → Store in collection")

    // Create enrichment pipeline
    val pipelineScript = UserFunction(
        label = PIPELINE_LABEL,
        name = "User Data Enrichment Pipeline",
        description = "Fetches external API data and stores enriched results",
        parameters = mapOf(
            "user_id" to ParameterDefinition(
                required = true,
                description = "User ID to enrich",
                paramType = "String"
            )
        ),
        functions = listOf(
            // Step 1: Fetch from external API with caching (30 min TTL)
            FunctionStageConfig.SWR(
                cache_key = "api:user:kt:{{user_id}}",
                ttl = JsonPrimitive("30m"),
                url = "https://jsonplaceholder.typicode.com/users/{{user_id}}",
                method = "GET",
                headers = null,
                body = null,
                timeout_seconds = null,
                output_field = "user_data",
                collection = null
            ),
            // Step 2: Store enriched data in collection (24 hour TTL)
            FunctionStageConfig.Insert(
                collection = ENRICHED_COLLECTION,
                record = buildJsonObject {
                    put("user_id", buildJsonObject {
                        put("type", JsonPrimitive("String"))
                        put("value", JsonPrimitive("{{user_id}}"))
                    })
                    put("source_data", buildJsonObject {
                        put("type", JsonPrimitive("Object"))
                        put("value", JsonPrimitive("{{user_data}}"))
                    })
                },
                bypass_ripple = false,
                ttl = 86400
            )
        ),
        tags = listOf("enrichment", "pipeline")
    )

    val pipelineScriptId = saveOrUpdate(client, pipelineScript)
    println("✓ Created enrichment pipeline: $PIPELINE_LABEL ($pipelineScriptId)")

    println("\nRunning pipeline:")
    val enrichResult = client.callFunction(PIPELINE_LABEL, mapOf("user_id" to JsonPrimitive("1")))
    println("  ✓ Data fetched from API (cached 30m)")
    println("  ✓ Enriched data stored in '$ENRICHED_COLLECTION' (TTL 24h)")
    println("  Pipeline returned ${enrichResult.records.size} records\n")

    pipelineScriptId
}

fun exampleDynamicTTL(client: EkoDBClient): String = runBlocking {
    println("\nExample 4: Dynamic TTL Configuration")
    println("─".repeat(80))
    println("TTL as parameter - supports duration strings, integers, ISO timestamps")

    // Create function with dynamic TTL
    val dynamicTTLScript = UserFunction(
        label = DYNAMIC_LABEL,
        name = "Flexible Cache TTL (Native SWR)",
        description = "Demonstrates parameterized TTL values",
        parameters = mapOf(
            "resource_id" to ParameterDefinition(
                required = true,
                description = "Resource to fetch",
                paramType = "String"
            ),
            "ttl" to ParameterDefinition(
                required = true,
                description = "Cache duration (e.g., '5m', '1h', '30s')",
                paramType = "String"
            )
        ),
        functions = listOf(
            FunctionStageConfig.SWR(
                cache_key = "resource:kt:{{resource_id}}",
                ttl = JsonPrimitive("{{ttl}}"),
                url = "https://jsonplaceholder.typicode.com/posts/{{resource_id}}",
                method = "GET",
                headers = null,
                body = null,
                timeout_seconds = null,
                output_field = null,
                collection = null
            )
        ),
        tags = listOf("dynamic")
    )

    val dynamicFuncId = saveOrUpdate(client, dynamicTTLScript)
    println("✓ Created dynamic TTL function: $DYNAMIC_LABEL ($dynamicFuncId)")

    // Test with different TTLs
    val ttlTests = listOf(
        "5m" to "5 minutes",
        "1h" to "1 hour",
        "30s" to "30 seconds"
    )

    for ((ttlValue, description) in ttlTests) {
        client.callFunction(
            DYNAMIC_LABEL,
            mapOf(
                "resource_id" to JsonPrimitive("test"),
                "ttl" to JsonPrimitive(ttlValue)
            )
        )
        println("  ✓ Cached with TTL: $ttlValue ($description)")
    }

    dynamicFuncId
}

suspend fun cleanupSwrFunctions(client: EkoDBClient) {
    println("\n🧹 Cleaning up...")
    val errors = mutableListOf<String>()
    var deletedFunctions = 0
    listOf(BASIC_LABEL, AUDIT_LABEL, PIPELINE_LABEL, DYNAMIC_LABEL).forEach { label ->
        try {
            val function = client.getFunction(label)
            val id = function.id ?: error("Function '$label' lookup returned no ID")
            client.deleteFunction(id)
            deletedFunctions += 1
        } catch (e: Exception) {
            if (!isNotFoundError(e)) {
                errors.add("function $label: ${e.message}")
            }
        }
    }
    listOf(
        "github:user:kt:torvalds",
        "product:kt:1",
        "api:user:kt:1",
        "resource:kt:test"
    ).forEach { key ->
        try {
            client.kvDelete(key)
        } catch (e: Exception) {
            if (!isNotFoundError(e)) {
                errors.add("KV key $key: ${e.message}")
            }
        }
    }
    listOf(AUDIT_COLLECTION, ENRICHED_COLLECTION).forEach { collection ->
        try {
            client.deleteCollection(collection)
        } catch (e: Exception) {
            if (!isNotFoundError(e)) {
                errors.add("collection $collection: ${e.message}")
            }
        }
    }
    check(errors.isEmpty()) { "Cleanup failed: ${errors.joinToString("; ")}" }
    println("✓ Deleted $deletedFunctions test functions and owned SWR collections")
}

fun main(): Unit = runBlocking {
    println("🚀 ekoDB Kotlin Client - Native SWR Function Examples\n")
    println("📋 Demonstrates:")
    println("   • Single-function SWR pattern (replaces 4-step pipeline)")
    println("   • Automatic cache checking, HTTP fetching, and cache setting")
    println("   • Built-in audit trail support")
    println("   • Duration string TTLs ('15m', '1h', '30s')")
    println("   • Multi-function pipeline integration")
    println("   • Dynamic TTL configuration\n")

    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"

    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()

    var failure: Throwable? = null
    try {
        cleanupSwrFunctions(client)
        // Run examples
        exampleBasicSWR(client)
        exampleAuditTrail(client)
        examplePipelineEnrichment(client)
        exampleDynamicTTL(client)

        // Summary
        println("\n${"=".repeat(80)}")
        println("✅ Key Benefits of Native SWR:")
        println("✅ Single function: Replaces 4-function cache-aside pattern")
        println("✅ Duration strings: Use '15m', '1h', '2h' instead of calculating seconds")
        println("✅ Built-in audit: Optional collection parameter for automatic logging")
        println("✅ Auto-enrichment: output_field populates params for downstream functions")
        println("✅ Transactional: Works correctly in both transactional and non-transactional contexts")
        println("✅ KV-optimized: Uses native KV store with proper TTL handling")

        println("\n=== Performance Comparison ===")
        println("Legacy Pattern: KvGet → If → HttpRequest → KvSet → Insert (5 functions)")
        println("Native SWR:     SWR → Insert (2 functions)")
        println("Result:         60% fewer functions, cleaner code, same behavior 🎯")
    } catch (error: Throwable) {
        println("❌ Error: ${error.message}")
        failure = error
    } finally {
        try {
            cleanupSwrFunctions(client)
        } catch (cleanupError: Throwable) {
            val primary = failure
            if (primary == null) failure = cleanupError else primary.addSuppressed(cleanupError)
        }
        try {
            client.close()
        } catch (cleanupError: Throwable) {
            val primary = failure
            if (primary == null) failure = cleanupError else primary.addSuppressed(cleanupError)
        }
    }

    failure?.let { throw it }

    println("\n✅ All examples completed!")
}

File: examples/rust/examples/client_swr_native.rs
//! Native SWR Function Example
//!
//! Demonstrates the new native SWR function in ekoDB that simplifies cache-aside patterns:
//! - Single function call handles cache lookup, HTTP fetch, and storage
//! - Built-in TTL support with duration strings ("15m", "1h", "2h")
//! - Optional audit trail collection for debugging
//! - Automatic enriched parameter population for downstream functions
//!
//! Compare this with client_swr_pattern.rs to see how much simpler the native function is!

use ekodb_client::{Client, FieldType, Function, ParameterDefinition, UserFunction};
use serde_json::json;
use std::collections::HashMap;
use std::time::Instant;

const BASIC_LABEL: &str = "github_user_native_rs";
const AUDIT_LABEL: &str = "product_swr_audit_rs";
const PIPELINE_LABEL: &str = "user_enrichment_pipeline_rs";
const DYNAMIC_LABEL: &str = "flexible_cache_rs";
const AUDIT_COLLECTION: &str = "swr_audit_trail_rs";
const ENRICHED_COLLECTION: &str = "enriched_users_rs";

fn is_not_found(error: &ekodb_client::Error) -> bool {
    matches!(error, ekodb_client::Error::Api { code: 404, .. })
        || error.to_string().to_lowercase().contains("not found")
}

async fn cleanup_owned(client: &Client) -> Result<(), Box<dyn std::error::Error>> {
    let mut errors = Vec::new();
    for label in [BASIC_LABEL, AUDIT_LABEL, PIPELINE_LABEL, DYNAMIC_LABEL] {
        match client.get_function(label).await {
            Ok(function) => match function.id {
                Some(id) => {
                    if let Err(error) = client.delete_function(&id).await {
                        if !is_not_found(&error) {
                            errors.push(format!("function {label}: {error}"));
                        }
                    }
                }
                None => errors.push(format!("function {label} lookup returned no ID")),
            },
            Err(error) => {
                if !is_not_found(&error) {
                    errors.push(format!("function {label} lookup: {error}"));
                }
            }
        }
    }
    for key in [
        "github:rs:torvalds",
        "product:rs:1",
        "user_api:rs:1",
        "resource:rs:test",
    ] {
        if let Err(error) = client.kv_delete(key).await {
            if !is_not_found(&error) {
                errors.push(format!("KV key {key}: {error}"));
            }
        }
    }
    for collection in [AUDIT_COLLECTION, ENRICHED_COLLECTION] {
        if let Err(error) = client.delete_collection(collection).await {
            if !is_not_found(&error) {
                errors.push(format!("collection {collection}: {error}"));
            }
        }
    }
    if errors.is_empty() {
        Ok(())
    } else {
        Err(format!("cleanup failed: {}", errors.join("; ")).into())
    }
}

/// Save a function idempotently: if the label already exists (HTTP 409),
/// update the existing definition instead, then return its id.
async fn save_or_update(
    client: &Client,
    function: UserFunction,
) -> Result<String, Box<dyn std::error::Error>> {
    let label = function.label.clone();
    match client.save_function(function.clone()).await {
        Ok(id) => Ok(id),
        Err(ekodb_client::Error::Api { code: 409, .. }) => {
            client.update_function(&label, function).await?;
            println!("ℹ️  Function '{}' already existed — updated instead", label);
            let existing = client.get_function(&label).await?;
            Ok(existing.id.unwrap_or(label))
        }
        Err(e) => Err(Box::new(e)),
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url =
        std::env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        std::env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    println!("=== ekoDB Native SWR Function ===\n");
    println!("This example shows the new simplified SWR function that replaces");
    println!("the manual FindById → If → HttpRequest → Insert pattern.\n");

    cleanup_owned(&client).await?;

    let run_result: Result<(), Box<dyn std::error::Error>> = async {
        // Example 1: Basic SWR with duration string TTL
        println!("Example 1: Basic GitHub User Cache with Native SWR");
        println!("─────────────────────────────────────────────────────\n");

        let basic_swr_script = UserFunction::new(BASIC_LABEL, "Native SWR GitHub User Cache")
            .with_description("Demonstrates native SWR function with cache-aside pattern")
            .with_parameter(
                ParameterDefinition::new("username")
                    .required()
                    .with_description("GitHub username to fetch"),
            )
            .with_function(Function::SWR {
                cache_key: "github:rs:{{username}}".to_string(),
                ttl: json!("15m"), // Duration string TTL
                url: "https://api.github.com/users/{{username}}".to_string(),
                method: "GET".to_string(),
                headers: Some(HashMap::from([(
                    "User-Agent".to_string(),
                    "ekoDB-Native-SWR/1.0".to_string(),
                )])),
                body: None,
                timeout_seconds: Some(10),
                output_field: Some("github_data".to_string()),
                collection: None, // No audit trail for this example
            })
            .with_tag("swr")
            .with_tag("native")
            .with_tag("github");

        let script_id = save_or_update(&client, basic_swr_script).await?;
        println!(
            "✓ Created native SWR script: {} ({})",
            BASIC_LABEL, script_id
        );

        // First call - cache miss
        println!("\nFirst call (cache miss - will fetch from GitHub API):");
        let start1 = Instant::now();
        let mut params1 = HashMap::new();
        params1.insert(
            "username".to_string(),
            FieldType::String("torvalds".to_string()),
        );

        let result1 = client.call_function(BASIC_LABEL, Some(params1)).await?;
        let duration1 = start1.elapsed();
        println!("  Response time: {}ms", duration1.as_millis());
        println!("  Records returned: {}", result1.records.len());
        println!("  ✓ Data fetched from API and cached with 15m TTL");

        // Second call - cache hit
        println!("\nSecond call (cache hit - instant from KV store):");
        let start2 = Instant::now();
        let mut params2 = HashMap::new();
        params2.insert(
            "username".to_string(),
            FieldType::String("torvalds".to_string()),
        );

        let _result2 = client.call_function(BASIC_LABEL, Some(params2)).await?;
        let duration2 = start2.elapsed();
        let speedup = if duration2.as_millis() > 0 {
            duration1.as_millis() as f64 / duration2.as_millis() as f64
        } else {
            0.0
        };
        println!("  Response time: {}ms", duration2.as_millis());
        println!("  Speedup: {:.1}x faster", speedup);
        println!("  ✓ Lightning fast cache hit\n");

        // Example 2: SWR with Audit Trail
        println!("\nExample 2: SWR with Audit Trail Collection");
        println!("─────────────────────────────────────────────────────\n");

        let audit_swr_script = UserFunction::new(AUDIT_LABEL, "Product Cache with Audit Trail")
            .with_description("Native SWR with audit collection for debugging")
            .with_parameter(
                ParameterDefinition::new("product_id")
                    .required()
                    .with_description("Product ID to fetch"),
            )
            .with_function(Function::SWR {
                cache_key: "product:rs:{{product_id}}".to_string(),
                ttl: json!("1h"), // 1 hour cache
                url: "https://dummyjson.com/products/{{product_id}}".to_string(),
                method: "GET".to_string(),
                headers: None,
                body: None,
                timeout_seconds: None,
                output_field: Some("product_data".to_string()),
                collection: Some(AUDIT_COLLECTION.to_string()), // Audit trail enabled
            })
            .with_tag("swr")
            .with_tag("audit");

        let audit_script_id = save_or_update(&client, audit_swr_script).await?;
        println!(
            "✓ Created SWR script with audit trail: {} ({})",
            AUDIT_LABEL, audit_script_id
        );

        println!("\nFetching product (will create audit trail entry):");
        let mut product_params = HashMap::new();
        product_params.insert("product_id".to_string(), FieldType::String("1".to_string()));

        let product_result = client
            .call_function(AUDIT_LABEL, Some(product_params))
            .await?;
        println!("  ✓ Product fetched and cached");
        println!(
            "  ✓ Audit record created in '{}' collection",
            AUDIT_COLLECTION
        );
        println!("  Records: {}\n", product_result.records.len());

        // Example 3: SWR with Data Enrichment Pipeline
        println!("\nExample 3: SWR in Multi-Function Pipeline");
        println!("─────────────────────────────────────────────────────\n");
        println!("Fetch external data → Process → Store in collection");

        let pipeline_script = UserFunction::new(PIPELINE_LABEL, "User Data Enrichment Pipeline")
            .with_description("Fetch from API, enrich with SWR, store processed result")
            .with_parameter(
                ParameterDefinition::new("user_id")
                    .required()
                    .with_description("User ID to enrich"),
            )
            // Step 1: Fetch user data with SWR caching
            .with_function(Function::SWR {
                cache_key: "user_api:rs:{{user_id}}".to_string(),
                ttl: json!("30m"),
                url: "https://jsonplaceholder.typicode.com/users/{{user_id}}".to_string(),
                method: "GET".to_string(),
                headers: None,
                body: None,
                timeout_seconds: None,
                output_field: Some("user_data".to_string()), // Populates enriched params
                collection: None,
            })
            // Step 2: Insert enriched data into permanent collection
            .with_function(Function::Insert {
                collection: ENRICHED_COLLECTION.to_string(),
                record: json!({
                    "user_id": {"type": "String", "value": "{{user_id}}"},
                    "source_data": {"type": "Object", "value": "{{user_data}}"} // Uses enriched param from SWR
                }),
                bypass_ripple: None,
                ttl: Some(json!("24h")), // Different TTL for permanent storage
            })
            .with_tag("enrichment")
            .with_tag("pipeline");

        let pipeline_script_id = save_or_update(&client, pipeline_script).await?;
        println!(
            "✓ Created enrichment pipeline: {} ({})",
            PIPELINE_LABEL, pipeline_script_id
        );

        println!("\nRunning pipeline:");
        let mut enrich_params = HashMap::new();
        enrich_params.insert("user_id".to_string(), FieldType::String("1".to_string()));

        let enrich_result = client
            .call_function(PIPELINE_LABEL, Some(enrich_params))
            .await?;
        println!("  ✓ Data fetched from API (cached 30m)");
        println!(
            "  ✓ Enriched data stored in '{}' (TTL 24h)",
            ENRICHED_COLLECTION
        );
        println!(
            "  Pipeline returned {} records\n",
            enrich_result.records.len()
        );

        // Example 4: Parameterized TTL
        println!("\nExample 4: Dynamic TTL Configuration");
        println!("─────────────────────────────────────────────────────\n");

        let dynamic_ttl_script =
            UserFunction::new(DYNAMIC_LABEL, "Flexible Cache with Dynamic TTL")
                .with_description("SWR function with user-controlled TTL")
                .with_parameter(
                    ParameterDefinition::new("resource_id")
                        .required()
                        .with_description("Resource identifier"),
                )
                .with_parameter(
                    ParameterDefinition::new("ttl")
                        .with_default(FieldType::String("10m".to_string()))
                        .with_description("Cache TTL (e.g., '5m', '1h', '2h')"),
                )
                .with_function(Function::SWR {
                    cache_key: "resource:rs:{{resource_id}}".to_string(),
                    ttl: json!("{{ttl}}"), // User-provided TTL
                    url: "https://jsonplaceholder.typicode.com/posts/{{resource_id}}".to_string(),
                    method: "GET".to_string(),
                    headers: None,
                    body: None,
                    timeout_seconds: None,
                    output_field: Some("resource".to_string()),
                    collection: None,
                })
                .with_tag("dynamic");

        let dynamic_script_id = save_or_update(&client, dynamic_ttl_script).await?;
        println!(
            "✓ Created dynamic TTL script: {} ({})",
            DYNAMIC_LABEL, dynamic_script_id
        );

        // Test with different TTLs
        let ttl_tests = vec![("5m", "5 minutes"), ("1h", "1 hour"), ("30s", "30 seconds")];

        for (ttl_value, description) in ttl_tests {
            let mut dynamic_params = HashMap::new();
            dynamic_params.insert(
                "resource_id".to_string(),
                FieldType::String("test".to_string()),
            );
            dynamic_params.insert("ttl".to_string(), FieldType::String(ttl_value.to_string()));

            let _ = client
                .call_function(DYNAMIC_LABEL, Some(dynamic_params))
                .await?;
            println!("  ✓ Cached with TTL: {} ({})", ttl_value, description);
        }

        println!("\n=== Key Benefits of Native SWR Function ===");
        println!("✅ Simpler: One function instead of 4 (FindById → If → HttpRequest → Insert)");
        println!("✅ Duration strings: Use '15m', '1h', '2h' instead of calculating seconds");
        println!("✅ Built-in audit: Optional collection parameter for automatic logging");
        println!("✅ Auto-enrichment: output_field populates params for downstream functions");
        println!(
        "✅ Transactional: Works correctly in both transactional and non-transactional contexts"
    );
        println!("✅ KV-optimized: Uses native KV store with proper TTL handling");

        println!("\n=== Performance Comparison ===");
        println!("Old pattern (manual):  ~4 function calls, complex script logic");
        println!("New pattern (native):  ~1 function call, handled by server");
        println!("Result: Simpler code, faster execution, easier maintenance\n");

        Ok(())
    }
    .await;

    // Cleanup
    println!("🧹 Cleaning up test data...");
    let cleanup_result = cleanup_owned(&client).await;
    match (run_result, cleanup_result) {
        (Err(primary), Err(cleanup)) => {
            return Err(format!("{primary}; cleanup also failed: {cleanup}").into());
        }
        (Err(primary), Ok(())) => return Err(primary),
        (Ok(()), Err(cleanup)) => return Err(cleanup),
        (Ok(()), Ok(())) => {}
    }
    println!("✓ Cleanup complete\n");

    println!("=== Summary ===");
    println!("The native SWR function provides a production-ready cache-aside pattern");
    println!("with minimal code and maximum flexibility. Use it for:");
    println!("  • API gateway caching");
    println!("  • External API aggregation");
    println!("  • Microservice response caching");
    println!("  • Edge computing patterns");
    println!("  • Real-time data enrichment pipelines\n");

    Ok(())
}
